src/PairedRouting.php

```diff
@@ -0,0 +1,901 @@
+				$wpdb->prepare(
+					"SELECT COUNT(*) FROM $wpdb->posts WHERE post_name = %s AND post_type = %s AND ID != %d LIMIT 1",
+					$alt_slug,
+					$post_type,
+					$post_id
+				)
+			);
+			$suffix++;
+		} while ( $slug_check );
+		$slug = $alt_slug;
+
+		return $slug;
+	}
+
+	/**
+	 * Add hooks based for AMP pages and other hooks for non-AMP pages.
+	 */
+	public function add_paired_request_hooks() {
+		if ( $this->has_endpoint() ) {
+			add_filter( 'old_slug_redirect_url', [ $this, 'maybe_add_paired_endpoint' ], 1000 );
+			add_filter( 'redirect_canonical', [ $this, 'maybe_add_paired_endpoint' ], 1000 );
+		} else {
+			add_action( 'wp_head', 'amp_add_amphtml_link' );
+		}
+	}
+
+	/**
+	 * Add notice to permalink settings screen for where to customize the paired URL structure.
+	 */
+	public function add_permalink_settings_notice() {
+		if ( 'options-permalink' !== get_current_screen()->id ) {
+			return;
+		}
+		?>
+		<div class="notice notice-info">
+			<p>
+				<?php
+				echo wp_kses(
+					sprintf(
+						/* translators: %s is the URL to the settings screen */
+						__( 'To customize the structure of the paired AMP URLs (given the site is not using the Standard template mode), go to the <a href="%s">Paired URL Structure</a> section on the AMP settings screen.', 'amp' ),
+						esc_url( admin_url( add_query_arg( 'page', AMP_Options_Manager::OPTION_NAME, 'admin.php' ) ) . '#paired-url-structure' )
+					),
+					[ 'a' => array_fill_keys( [ 'href' ], true ) ]
+				);
+				?>
+			</p>
+		</div>
+		<?php
+	}
+
+	/**
+	 * Determine whether permalinks are being used.
+	 *
+	 * @return bool If permalinks are enabled.
+	 */
+	public function is_using_permalinks() {
+		return ! empty( get_option( 'permalink_structure' ) );
+	}
+
+	/**
+	 * Add default option.
+	 *
+	 * @param array $defaults Default options.
+	 * @param array $options  Current options.
+	 * @return array Defaults.
+	 */
+	public function filter_default_options( $defaults, $options ) {
+		$value = Option::PAIRED_URL_STRUCTURE_QUERY_VAR;
+
+		if (
+			isset( $options[ Option::VERSION ], $options[ Option::THEME_SUPPORT ], $options[ Option::READER_THEME ] )
+			&&
+			version_compare( $options[ Option::VERSION ], '2.1', '<' )
+		) {
+			if (
+				AMP_Theme_Support::READER_MODE_SLUG === $options[ Option::THEME_SUPPORT ]
+				&&
+				ReaderThemes::DEFAULT_READER_THEME === $options[ Option::READER_THEME ]
+				&&
+				$this->is_using_permalinks()
+			) {
+				$value = Option::PAIRED_URL_STRUCTURE_LEGACY_READER;
+			} elseif ( AMP_Theme_Support::STANDARD_MODE_SLUG !== $options[ Option::THEME_SUPPORT ] ) {
+				$value = Option::PAIRED_URL_STRUCTURE_LEGACY_TRANSITIONAL;
+			}
+		}
+
+		$defaults[ Option::PAIRED_URL_STRUCTURE ] = $value;
+
+		return $defaults;
+	}
+
+	/**
+	 * Sanitize options.
+	 *
+	 * @todo This is redundant with the enum defined in the schema.
```

Comment:
How so? Wouldn't this ensure that the option is sanitized correctly when the `amp_options` option is being updated in a non-REST manner?

Reply:
Correct, but as I recall we exclusively use the REST API to update settings. Nevertheless, I'll change it as follows:

```suggestion
	 * Note that in a REST API context this is redundant with the enum defined in the schema.
```